aospy/test/test_utils_times.py

```diff
@@ -421,6 +315,94 @@ def test_assert_has_data_for_time():
+    '360_day': cftime.Datetime360Day,
+    'julian': cftime.DatetimeJulian,
+    'all_leap': cftime.DatetimeAllLeap,
+    '366_day': cftime.DatetimeAllLeap,
+    'gregorian': cftime.DatetimeGregorian,
+    'proleptic_gregorian': cftime.DatetimeProlepticGregorian
+}
+
+
+@pytest.mark.parametrize(['calendar', 'date_type'],
+                         list(CFTIME_DATE_TYPES.items()))
+def test_assert_has_data_for_time_cftime_datetimes(calendar, date_type):
+    time_bounds = np.array([[0, 2], [2, 4], [4, 6]])
+    nv = np.array([0, 1])
+    time = np.array([1, 3, 5])
+    data = np.zeros((3))
+    var_name = 'a'
+    ds = xr.DataArray(data,
+                      coords=[time],
+                      dims=[TIME_STR],
+                      name=var_name).to_dataset()
+    ds[TIME_BOUNDS_STR] = xr.DataArray(time_bounds,
+                                       coords=[time, nv],
+                                       dims=[TIME_STR, BOUNDS_STR],
+                                       name=TIME_BOUNDS_STR)
+    units_str = 'days since 0002-01-02 00:00:00'
+    ds[TIME_STR].attrs['units'] = units_str
+    ds[TIME_STR].attrs['calendar'] = calendar
+    ds = ensure_time_avg_has_cf_metadata(ds)
+    ds = set_grid_attrs_as_coords(ds)
+    with xr.set_options(enable_cftimeindex=True):
+        ds = xr.decode_cf(ds)
+    da = ds[var_name]
+
+    start_date = date_type(2, 1, 2)
+    end_date = date_type(2, 1, 8)
+    _assert_has_data_for_time(da, start_date, end_date)
+
+    start_date_bad = date_type(2, 1, 1)
+    end_date_bad = date_type(2, 1, 9)
+
+    with pytest.raises(AssertionError):
+        _assert_has_data_for_time(da, start_date_bad, end_date)
+
+    with pytest.raises(AssertionError):
+        _assert_has_data_for_time(da, start_date, end_date_bad)
+
+    with pytest.raises(AssertionError):
+        _assert_has_data_for_time(da, start_date_bad, end_date_bad)
+
+
+def test_assert_has_data_for_time_str_input():
+    time_bounds = np.array([[0, 31], [31, 59], [59, 90]])
+    nv = np.array([0, 1])
+    time = np.array([15, 46, 74])
+    data = np.zeros((3))
+    var_name = 'a'
+    ds = xr.DataArray(data,
+                      coords=[time],
+                      dims=[TIME_STR],
+                      name=var_name).to_dataset()
+    ds[TIME_BOUNDS_STR] = xr.DataArray(time_bounds,
+                                       coords=[time, nv],
+                                       dims=[TIME_STR, BOUNDS_STR],
+                                       name=TIME_BOUNDS_STR)
+    units_str = 'days since 2000-01-01 00:00:00'
+    ds[TIME_STR].attrs['units'] = units_str
+    ds = ensure_time_avg_has_cf_metadata(ds)
+    ds = set_grid_attrs_as_coords(ds)
+    ds = xr.decode_cf(ds)
+    da = ds[var_name]
+
+    start_date = '2000-01-01'
+    end_date = '2000-03-31'
+    _assert_has_data_for_time(da, start_date, end_date)
+
+    start_date_bad = '1999-12-31'
+    end_date_bad = '2000-04-01'
+
+    # With strings these checks are disabled
```

Comment:
Why this difference in behavior between strings and datetime-like types?

Reply:
Partial datetime strings are somewhat complicated.  It involves a fair bit of code (and tests) to fully support parsing them and interpreting what they mean as bounds.  While much of that code exists in xarray (see [here in `cftimeindex.py`](https://github.com/pydata/xarray/blob/9c8005937556211a8bf28a946744da3768846c5a/xarray/coding/cftimeindex.py#L12-L100)), it is private API.  At least to me, this sanity check in aospy is not important enough to merit suggesting that xarray support `parse_iso8601`, `_parse_iso8601_with_reso `, and `_parsed_string_to_bounds ` as public API.  While they are used by xarray, these functions are outside its scope (i.e. xarray is not a date-parsing library).

If we want to support this check with strings, it then leaves us with two options:
- Port all of that code (and tests) to aospy.
- Use those functions from xarray, despite the fact that they are private API.

In your opinion, are either of those are worth it?

Reply:
Thanks, that's helpful. Is this the same reason for the different handling of strings vs. other types elsewhere in the code?  Frankly, it does concern me a little that, for a user, behavior could be different whether they supply dates as strings or as a datetime-like type.

Putting the above question another way: let's suppose we went ahead and ported that code or imported it.  Would that enable you to, as soon as dates are read in as strings, convert them to a datetime-like type?  And, if yes, would that then remove the need for separate logic?  Purely from a design perspective, to me the optimal solution would be to create an "adapter" (something that takes strings or any of the supported datetime-like types and outputs a datetime object), apply it right away, and then have the remainder of the pipeline ingest/output a single type.  But from the technical/implementation perspective, perhaps there are good reasons that would argue against this that I'm still not appreciating.

Hope that makes sense!

Reply:
> Would that enable you to, as soon as dates are read in as strings, convert them to a datetime-like type?

I'm pretty reluctant to do that :)  The beauty of using strings is that we don't need to modify them at all!  We can just pass them directly as slice bounds when we call `sel` to subset a Dataset in time, regardless of whether the time coordinate is indexed by a DatetimeIndex or CFTimeIndex.  For the most part that is why the logic is different for them throughout what I've implemented here.  If it weren't for `_assert_has_data_for_time`, strings would be significantly easier to deal with than date objects (because we don't need to worry about their type).

Let me think about things a little more.  There may be a solution that doesn't involve porting significant amounts of existing code or using private API.

Reply:
OK, that makes sense to me.  In that case, what about, just within `_assert_has_data_for_time`, casting the string to one of the datetime objects?  Sorry if I'm still being naive here.  If not, at least maybe we should add a note in the docs and/or a logging message that string dates don't undergo this check?

Ultimately I trust your judgment on this one and don't want to bog you down on this point.  I don't think at all that the current implementation will incur a lot of technical debt (quite the opposite in fact).  So let's go with whatever you decide is best at this point.

Reply:
> In that case, what about, just within _assert_has_data_for_time, casting the string to one of the datetime objects?

Indeed, this is exactly what I had in mind when discussing the options here: https://github.com/spencerahill/aospy/pull/273#discussion_r191439050.  

> Ultimately I trust your judgment on this one and don't want to bog you down on this point.

No worries!  I understand where you are coming from.  In a perfect world, I agree it would be nice to support this check.  Would it help you understand where I'm coming from if I provided an explicit example for why casting string inputs as dates is difficult and something that might best be left for upstream libraries?

Reply:
>Indeed, this is exactly what I had in mind when discussing the options here: #273 (comment).

Sorry, right of course.  So that probably is overkill; I think a warning/docs note would be fine.

>Would it help you understand where I'm coming from if I provided an explicit example for why casting string inputs as dates is difficult and something that might best be left for upstream libraries?

An example would definitely help, but that might be overkill just for my behalf.  On the other hand, it occurred to me: would this functionality be appropriate in cftime?  In which case maybe a clean example would be good as part of an Issue on it in that repo.

Reply:
> An example would definitely help

For sure!  Let me provide some context here.

Ideally we would like to allow users to specify partial-datetime strings as time bounds for their calculations.  Partial-datetime strings are a very convenient shorthand; they allow you to select data between the earliest datetime possible for the lower bound, all the way to the greatest datetime possible on the upper bound.

For instance, if you wanted to compute things for data between year 2000 and 2010 (inclusive), you could just specify a date range of `('2000', '2010')`.  This is in contrast to using explicit datetimes, which would look like `(datetime(2000, 1, 1), datetime(2010, 12, 31, 23, 59, 59, 999999))`.  You can continue refining your interval by filling in more of the bounds.  So for instance if you wanted to select data between February, 2000 and February 2010, you could use `('2000-02', '2010-02')` and this would map to `(datetime(2000, 2, 1), datetime(2010, 2, 28, 23, 59, 59, 999999))`.  

For this reason, it is not a simple matter of "casting" a string to a particular datetime.  The datetime you transform it to matters based on the context.  For instance if `'2000-02'` is the lower bound, then it maps to `datetime(2000, 2, 1)`, while if it is the upper bound it maps to `datetime(2000, 2, 29, 23, 59, 59, 999999)`.  pandas and now xarray have logic for doing this for numpy and cftime dates, so it sort of feels heavy-handed to re-implement it in aospy.

> So that probably is overkill; I think a warning/docs note would be fine.

Sounds good.  I'll mull this over for the next few days.  If I can't think of a simple solution, I'll go forward with the warning/docs note.

Reply:
Ok that makes total sense.  Thanks.

>I'll mull this over for the next few days. If I can't think of a simple solution, I'll go forward with the warning/docs note.

Sounds good.  My only remaining question is re: if this functionality should/could go in cftime eventually.  Perhaps Jeremy's comment on [this thread](https://groups.google.com/forum/#!topic/xarray/1-mzcXQNVII) on the xarray mailing list is relevant?

Reply:
> My only remaining question is re: if this functionality should/could go in cftime eventually. Perhaps Jeremy's comment on this thread on the xarray mailing list is relevant?

I agree with Jeremy and think a cftime version of `datetime.strptime` would be great; it could help in this case.

Reply:
Great.  I opened an issue to start that discussion: https://github.com/Unidata/cftime/issues/51

Who knows if/when that will progress though, so let's not wait up for it for this PR.